src/uiprotect/data/types.py

```diff
@@ -2,20 +2,42 @@
 
 import enum
 from collections.abc import Callable, Coroutine
-from functools import cache
-from typing import Any, Literal, Optional, TypeVar, Union
+from functools import cache, lru_cache
+from typing import Annotated, Any, Literal, Optional, TypeVar, Union
 
 from packaging.version import Version as BaseVersion
-from pydantic.v1 import BaseModel, ConstrainedInt
-from pydantic.v1.color import Color as BaseColor
-from pydantic.v1.types import ConstrainedFloat, ConstrainedStr
+from pydantic import BaseModel, Field
+from pydantic.fields import FieldInfo
+from pydantic.types import StringConstraints
+from pydantic.v1.config import BaseConfig as BaseConfigV1
+from pydantic.v1.fields import SHAPE_DICT as SHAPE_DICT_V1  # noqa: F401
+from pydantic.v1.fields import SHAPE_LIST as SHAPE_LIST_V1  # noqa: F401
+from pydantic.v1.fields import SHAPE_SET as SHAPE_SET_V1  # noqa: F401
+from pydantic.v1.fields import ModelField as ModelFieldV1
+from pydantic_extra_types.color import Color as BaseColor
 
 from .._compat import cached_property
 
 KT = TypeVar("KT")
 VT = TypeVar("VT")
 
 
+class _BaseConfigV1(BaseConfigV1):
+    arbitrary_types_allowed = True
+    validate_assignment = True
+
+
+@lru_cache
+def extract_type_shape(field: FieldInfo) -> tuple[Any, int]:
+    """Extract the type from a type hint."""
+    type_ = field.annotation
+    assert type_ is not None
+    v1_field = ModelFieldV1(
+        name="", type_=type_, class_validators=None, model_config=_BaseConfigV1
+    )
+    return v1_field.type_, v1_field.shape
+
```

Comment:
pydantic dropped `type_` and `shape` in v2, but since we need it, this is a workaround to get it